File: app/models/recipe.py
"""Recipe data models using Pydantic."""

from pydantic import BaseModel
from typing import List, Optional


class CacheInfo(BaseModel):
    """Cache performance information."""
    hit: bool                    # True if cache hit, False if cache miss
    response_time_ms: float     # Response time in milliseconds
    source: str                 # "cache" or "api"


class Recipe(BaseModel):
    """Complete recipe model with all fields."""
    id: Optional[int] = None
    title: str
    ingredients: List[str]
    steps: List[str]
    prepTime: str
    cookTime: str
    difficulty: str
    cuisine: str
    source: str = "internal"  # Default to internal for existing recipes


class RecipeRequest(BaseModel):
    """Request model for creating/updating recipes (without ID)."""
    title: str
    ingredients: List[str]
    steps: List[str]
    prepTime: str
    cookTime: str
    difficulty: str
    cuisine: str


class RecipeResponse(BaseModel):
    """Response model (Recipe with ID)."""
    id: int
    title: str
    ingredients: List[str]
    steps: List[str]
    prepTime: str
    cookTime: str
    difficulty: str
    cuisine: str
    source: str
    cache_info: Optional[CacheInfo] = None

File: frontend/src/lib/types/recipe.ts
/**
 * TypeScript types matching the Python Pydantic models from the Recipe Discovery API
 */

export interface Recipe {
	id?: number;
	title: string;
	ingredients: string[];
	steps: string[];
	prepTime?: string | null;  // Optional for external recipes
	cookTime?: string | null;  // Optional for external recipes
	difficulty?: string | null;  // Optional for external recipes
	cuisine?: string | null;  // Optional for external recipes
	source?: string;
}

export interface RecipeRequest {
	title: string;
	ingredients: string[];
	steps: string[];
	prepTime: string;
	cookTime: string;
	difficulty: string;
	cuisine: string;
}

export interface CacheInfo {
	hit: boolean;               // True if cache hit, False if cache miss
	response_time_ms: number;   // Response time in milliseconds
	source: string;             // "cache" or "api"
}

export interface RecipeResponse {
	id: number;
	title: string;
	ingredients: string[];
	steps: string[];
	prepTime?: string | null;  // Optional for external recipes
	cookTime?: string | null;  // Optional for external recipes
	difficulty?: string | null;  // Optional for external recipes
	cuisine?: string | null;  // Optional for external recipes
	source: string;
	cache_info?: CacheInfo;     // Cache performance data
}

export interface SearchResultsResponse {
	recipes: RecipeResponse[];
	mealdb_cache_info?: CacheInfo;
}

export interface ApiError {
	detail: string;
}
